### DISCUSSION TITLE
SyntaxTree: How to read full `ValueText` of a multi-line Raw Literal String (or: why is it different from a multi-line verbatim string?)

### DISCUSSION THREAD

**Discussion by BrunoJuchli** [dotnet/roslyn#72271] [Q&A] [answered]
**Edit: It looks like I made a mistake in passing my "source code" to the parser. And I got confused .. so my bad ;-)**

## Question
How can I use a parsed `SyntaxTree` to get the whole value text of a multi-line raw literal string?
Why is it different from how we can retrieve the value text of a multi-line verbatim string?

## How I Parse my Code

Using `Microsoft.CodeAnalysis` v`4.8.0` (or `4.9.0-3.final`), with:

```C#
CSharpSyntaxTree
    .ParseText(
       "source code - I'll give examples below",
       path: "dummy",
       options: CSharpParseOptions
           .Default
           .WithLanguageVersion(LanguageVersion.Latest));
```

### Verbatim multi-line string

with the following source text:

```C#
S[@"a verbatim
    string text"];
```

Just to be sure that I don't make any mistakes, this is how I define it in my c# code:
```C#
@"
S[""""""a raw literal
     multiline string
     """"""];"
```

--> since i use a @-verbatim multiline string to declare the source, I need `""` to escape a single `"`.

#### Result

`ElementAccessExpressionSyntax.ArgumentList.Arguments` contains **a single** `ArgumentSyntax` with an `.Expression` of type `LiteralExpressionSyntax` with `.Token.ValueText` = `A verbatim\r\n    string text`.

=> I can get the string value via  `((LiteralExpressionSyntax)argument.Expression).Token.ValueText`.

### Multiline Raw Literal String

```C#
S["""a raw literal 
       multiline string
       """];
```

#### Result

`ElementAccessExpressionSyntax.ArgumentList.Arguments` contains **4** `ArgumentSyntax`s, where `.Expression` is:
- 0 - `LiteralExpressionSyntax` with `.Token.ValueText = `a raw literal`.
- 1 - `IdentifierNameSyntax` with `Identifier.Text` = `multiline`
- 2 - `PredefinedTypeSyntax` with `.Keyword.ValueText` = `string`
- 3 - `LiteralExpressionSyntax` with `.Expression.Token.ValueText = `];`.

Note that `ElementAccessExpressionSyntax.ToString()` returns `S["""a raw literal\r\n     multiline string\r\n     """];`;

**Question**: how am I supposed to extract `a raw literal multiline string` from this?
Is this broken / buggy / support for multi-line raw literal strings is missing?

**Comment by BrunoJuchli** [comment]
Ok so it looks like I just got confused and didn't properly specify the multiline raw literal.

Here's how it works:

```C#
@"
S[""""""
     a raw literal
     multiline string
     """"""];"
```

resulting in a single argument with `.Expression.Token.ValueText` = `a raw literal\r\nmultiline string`.

**Comment by CyrusNajmabadi** [comment, marked as answer]
Your raw string literal is incorrect.  It cannot cross multiple lines without starting with a new line after the initial quotes.

**Reply by BrunoJuchli** [reply]
Thanks, you're right, I got confused there :-/

**Comment by CyrusNajmabadi** [comment]
> ElementAccessExpressionSyntax.ArgumentList.Arguments contains 4 ArgumentSyntaxs, where .Expression is:

Check the diagnostics on the syntax tree instance you get back.  It should indicate the problem.
